sdk/cosmos/azure_data_cosmos/src/location_cache.rs

```diff
@@ -0,0 +1,570 @@
+        read_locations: HashMap<String, String>,
+    ) -> Result<(), &'static str> {
+        // Separate write locations into appropriate hashmap and list
+        if !write_locations.is_empty() {
+            let (available_write_endpoints_by_location, available_write_locations) =
+                self.get_endpoints_by_location(write_locations);
+            self.locations_info.available_write_endpoints_by_location =
+                available_write_endpoints_by_location;
+            self.locations_info.available_write_locations = available_write_locations;
+        }
+
+        // Separate read locations into appropriate hashmap and list
+        if !read_locations.is_empty() {
+            let (available_read_endpoints_by_location, available_read_locations) =
+                self.get_endpoints_by_location(read_locations);
+            self.locations_info.available_read_endpoints_by_location =
+                available_read_endpoints_by_location;
+            self.locations_info.available_read_locations = available_read_locations;
+        }
+
+        self.refresh_endpoints();
+
+        Ok(())
+    }
+
+    pub fn mark_endpoint_unavailable(&mut self, endpoint: &str, operation: RequestOperation) {
+        let now = std::time::SystemTime::now();
+
+        {
+            let mut location_unavailability_info_map =
+                self.location_unavailability_info_map.lock().unwrap();
+
+            if let Some(info) = location_unavailability_info_map.get_mut(endpoint) {
+                if info.unavailable_operation == operation
+                    || info.unavailable_operation == RequestOperation::All
+                {
+                    // If the endpoint is already marked as unavailable, update last_check_time
+                    info.last_check_time = now;
+                } else {
+                    // If endpoint marked unavailable but with a different operation, update operation type to All
+                    info.last_check_time = now;
+                    info.unavailable_operation = RequestOperation::All;
+                }
+            } else {
+                // If the endpoint is not in the map, insert it
+                location_unavailability_info_map.insert(
+                    endpoint.to_string(),
+                    LocationUnavailabilityInfo {
+                        last_check_time: now,
+                        unavailable_operation: operation,
+                    },
+                );
+            }
+        }
+
+        self.refresh_endpoints();
+    }
+
+    pub fn is_endpoint_unavailable(&self, endpoint: &str, operation: RequestOperation) -> bool {
+        let location_unavailability_info_map =
+            self.location_unavailability_info_map.lock().unwrap();
+        if let Some(info) = location_unavailability_info_map.get(endpoint) {
+            // Checks if endpoint is unavailable for the given operation
+            let operation_type = info.unavailable_operation == operation;
```

Comment:
nit: A more descriptive name for this boolean would be `operation_type_available`